swift/apple/FirezoneKit/Sources/FirezoneKit/Models/FirezoneId.swift

```diff
@@ -0,0 +1,124 @@
+//
+//  FirezoneId.swift
+//  (c) 2024 Firezone, Inc.
+//  LICENSE: Apache-2.0
+//
+//  Convenience wrapper for working with our firezone-id stored in the Keychain.
+
+import Foundation
+
+public struct FirezoneId {
+  private static let query: [CFString: Any] = [
+    kSecAttrLabel: "Firezone id",
+    kSecAttrAccount: "2",
+    kSecAttrService: AppInfoPlistConstants.appGroupId,
+    kSecAttrDescription: "Firezone device id",
+  ]
+
+  private var uuid: UUID
+
+  public init(_ uuid: UUID? = nil) {
+    self.uuid = uuid ?? UUID()
+  }
+
+  // Upsert the firezone-id to the Keychain
+  public func save() async throws {
+    guard await Keychain.shared.search(query: FirezoneId.query) == nil
+    else {
+      let query = FirezoneId.query.merging([
+        kSecClass: kSecClassGenericPassword
+      ]) { (_, new) in new }
+      return try await Keychain.shared.update(
+        query: query,
+        attributesToUpdate: [kSecValueData: uuid.toData()]
+      )
+    }
+
+    let query = FirezoneId.query.merging([
+      kSecClass: kSecClassGenericPassword,
+      kSecValueData: uuid.toData()
+    ]) { (_, new) in new }
+
+    try await Keychain.shared.add(query: query)
+  }
+
+  // Attempt to load the firezone-id from the Keychain
+  public static func load() async throws -> FirezoneId? {
+    guard let idRef = await Keychain.shared.search(query: query)
+    else { return nil }
+
+    guard let data = await Keychain.shared.load(persistentRef: idRef)
+    else { return nil }
+
+    guard data.count == UUID.sizeInBytes
+    else {
+      fatalError("Firezone ID loaded from keychain must be exactly \(UUID.sizeInBytes) bytes")
+    }
+
+    let uuid = UUID(fromData: data)
+    return FirezoneId(uuid)
+  }
+
+  // Prior to 1.4.0, our firezone-id was saved in a file. Starting with 1.4.0,
+  // the macOS client uses a system extension, which makes sharing folders with
+  // the app cumbersome, so we moved to using the keychain for this due to its
+  // better ergonomics.
+  //
+  // Can be refactored to remove the file check once all clients >= 1.4.0
+  public static func createIfMissing() async throws {
```

Comment:
This function does more than the name suggests. Why not split the migration code out into a separate function and call that explicitly on start-up? Then it can be more easily deleted later.

Reply:
Split the migrate out into a `migrate()`

Reply:
Split the migrate out into a `migrate()`